### ISSUE TITLE
Request for Advice on Improving Unitree B2 Locomotion Policy for Locomotion-B2-Rough-Blind

### ISSUE THREAD

**Issue by Siwakonrome** [iit-DLSLab/basic-locomotion-dls-isaaclab#50]
I have been testing the Locomotion-B2-Rough-Blind task using the default environment and configuration from the repository.

The training command I used is:
```python
python scripts/rsl_rl/train.py \
    --task=Locomotion-B2-Rough-Blind \
    --num_envs=4096 \
    --headless \
    --logger=tensorboard
```
The training settings were: max_iterations = 8000, num_envs = 4096
After the training was completed, I found that the robot was able to stand, but not very well. More importantly, the robot was not able to move properly according to the task velocity command.

I would like to ask whether I need to adjust any additional configurations, such as the actuator Kp and Kd values.

Previously, I tested joint position control on the real Unitree B2 robot through the ROS2 /lowcmd topic, and I found that the Unitree example code uses the following values: (float kp_ = 1000.0;, float kd_ = 10.0;)

Reference:
https://github.com/unitreerobotics/unitree_ros2/blob/master/example/src/src/b2/b2_stand_example.cpp

My question is:
Do I need to modify the Kp and Kd values in the simulation or training configuration to match the values used in b2_stand_example.cpp?
and In order to make Locomotion-B2-Rough-Blind walk properly in MuJoCo, what configurations should I adjust or tune?

I would greatly appreciate your guidance and recommendations.

---
<img width="400" height="225" alt="Image" src="https://github.com/user-attachments/assets/497603bb-6d57-407c-b379-d8ccb64b2e00" />
---

## Locomotion-B2-Rough-Blind — reference tables

Primary sources: `source/basic_locomotion_isaaclab/.../b2_env_cfg.py`, `locomotion_env.py`, `assets/b2_asset.py`, `agents/rsl_rl_ppo_cfg.py`, `deploy/config.py`, `deploy/play_mujoco.py`, `deploy/locomotion_policy_wrapper.py`.

### 1) Task structure

| Item | Value |
|------|--------|
| Gym id | `Locomotion-B2-Rough-Blind` |
| Env class | `LocomotionEnv` (`DirectRLEnv`) |
| Env config | `B2RoughBlindEnvCfg` (extends `B2FlatEnvCfg`) |
| Agent config | `RoughPPORunnerCfg` (`rsl_rl_ppo_cfg:RoughPPORunnerCfg`) |
| Log experiment folder | `logs/rsl_rl/rough_direct/<run_timestamp>/` |

### 2) Environment & simulation

| Item | Value |
|------|--------|
| `episode_length_s` | 20 |
| `sim.dt` | 1/200 s (200 Hz physics) |
| `decimation` | 4 → policy / reward step **0.02 s** (50 Hz) |
| Default `scene.num_envs` | 4096 (`--num_envs` overrides at train time) |
| `env_spacing` | 4.0 |
| Terrain (Rough-Blind) | `terrain_type="generator"`, **custom** `ROUGH_TERRAINS_CFG` in `B2RoughBlindEnvCfg`: mixed sub-terrains (flat, boxes, star, rough HF, slopes, stairs), **curriculum=False**, patch **8×8** m, grid **10×20**, `use_cache=False` |
| Flat baseline (`B2FlatEnvCfg`) | Plane terrain |
| Sensors (shared) | Contact sensor on robot; **height_scanner** RayCaster on base (reward / terrain pitch — **not** policy height-map for Blind); IMU prim defined |
| Action / observation noise | Gaussian + bias on action (`std` 0.05, bias 0.015) and observation (`std` 0.002, bias 0.0001) |
| Gait / tracking targets | `desired_base_height` **0.485**, `desired_feet_height` **0.05**, `desired_step_freq` **1.4** Hz, `desired_duty_factor` **0.65**, phase offsets FL/FR/RL/RR **0 / 0.5 / 0.5 / 0** |

### 3) Robot & actuator (Isaac `b2_asset.py`)

| Group | `stiffness` (Kp-like) | `damping` (Kd-like) | Effort limit | Vel limit |
|-------|------------------------|----------------------|--------------|-----------|
| Hip / Thigh | **25** | **2** | 200 N·m | 23 rad/s |
| Calf | **25** | **2** | 300 N·m | 14 rad/s |
| Other actuator params | `friction_static` 0.2, `friction_dynamic` 0.6, `armature` 0.01, `activation_vel` 0.1 | | | |
| Spawn pose | z = 0.5 m; thigh **0.9** rad, calf **-1.8** rad | | | |
| `ArticulationRootPropertiesCfg` | self-collisions **on**, pos solver iters **4**, vel solver iters **0** | | | |

### 4) Domain randomization (`EventCfg` in `b2_env_cfg.py`)

| Event | Mode | Summary |
|-------|------|---------|
| `physics_material` | startup | Randomize body friction **0.2–1.25**, restitution **0–0.1**, 64 buckets |
| `add_base_mass` | startup | Add mass on **base** **-5 … +5** kg |
| `scale_all_link_masses` | startup | Scale all link masses **0.9–1.1** |
| `base_external_force_torque` | reset | Force / torque on base **±5** |
| `scale_all_joint_friction_model` | startup | Scale joint friction **0.2–2.0** |
| `scale_all_joint_armature_model` | startup | Scale armature **0–1** |
| `actuator_gains` | reset | Add to stiffness **±5**, damping **±1** (uniform) |
| `push_robot` | interval **10–15 s** | Random velocity perturbation **±0.5** (x,y,z, roll, pitch, yaw) |

### 5) Action / observation

| Item | Value |
|------|--------|
| `action_space` | 12 |
| `desired_clip_actions` | ±3.0 |
| `action_scale` | 0.5 → joint position target = default + `action_scale` × (filtered action) |
| `use_filter_actions` | True (α = 0.8 in `locomotion_env`) |
| Single-step obs dim (before history) | **52** = **48** base + **4** clock (`use_clock_signal=True`) |
| `use_observation_history` / `history_length` | True / **5** → policy vector dim **260** |
| Policy observation content | Base lin vel (or IMU if enabled), ang vel, projected gravity, velocity commands, joint pos error, joint vel, previous actions, gait clock |
| `use_vision` | **False** for Rough-Blind (no height-map channel in policy obs) |

### 6) Rewards — terms and scales (`B2FlatEnvCfg` → `locomotion_env._get_rewards`)

Each logged term is **(metric) × scale × `step_dt`**; total reward per step is the **sum** of all terms. **`step_dt` = `decimation` × `sim.dt` = 0.02 s.**

| Reward key (log name) | Scale from cfg | Notes |
|-------------------------|----------------|--------|
| `track_height_exp` | `height_reward_scale` **1.0** | |
| `track_lin_vel_xy_exp` | `lin_vel_reward_scale` **2.0** | |
| `track_lin_vel_z_l2` | `z_vel_reward_scale` **-2.0** | |
| `track_orientation_l2` | `orientation_reward_scale` **-5.0** | |
| `track_ang_vel_xy_l2` | `ang_vel_reward_scale` **-0.25** | |
| `track_ang_vel_z_exp` | `yaw_rate_reward_scale` **0.5** | |
| `undesired_contacts` | `undersired_contact_reward_scale` **-1.0** | |
| `action_rate_l2` | `action_rate_reward_scale` **-0.01** | |
| `action_smoothness_l2` | `action_smoothness_reward_scale` **-0.001** | |
| `joints_hip_pos_l2` | **-0.1** | |
| `joints_thigh_pos_l2` | **-0.1** | |
| `joints_calf_pos_l2` | **-0.001** | |
| `joints_acc_l2` | **-2.5e-7** | |
| `joints_torques_l2` | **-2.5e-6** | |
| `joints_energy_l1` | **-1e-4** | |
| `feet_air_time` | **0** (`feet_air_time_reward_scale`) | Disabled |
| `feet_height_clearance` | **0** | Disabled |
| `feet_height_clearance_periodic` | **0.25** | |
| `feet_height_clearance_mujoco` | **0** | Disabled |
| `feet_height_clearance_mujoco_periodic` | **0** | Disabled |
| `feet_slide` | **0** | Disabled |
| `feet_contact_suggestion` | **0.25** | |
| `feet_to_base_distance_l2` | **0** | Disabled |
| `feet_to_hip_distance_l2` | **1.5** | Uses `desired_hip_offset` **0.12** |
| `feet_vertical_surface_contacts` | **-0.25** | |

**Total reward & “max reward”**

| Topic | Detail |
|-------|--------|
| Per-step total | `reward = sum(all terms)` (each term already includes its scale and `step_dt`). |
| Fixed theoretical maximum | **Not defined** in code — many terms are unbounded or scenario-dependent (terrain, contacts). |
| What “good walking” looks like | Compare **mean episode return** and **`Episode_Reward/*`** in TensorBoard across training / checkpoints; rising **`track_lin_vel_xy_exp`**, **`track_ang_vel_z_exp`**, episode length, and falling bad contacts / early termination indicate improvement — not a single numeric cap. |

### 7) Termination

| Signal | Condition |
|--------|-----------|
| **Termination (`died`)** | Contact force on **base** or **hip** bodies (undesired ground contact) |
| **Timeout** | Episode length reaches limit |

### 8) PPO (`RoughPPORunnerCfg`) — summary

| Field | Value |
|-------|--------|
| `experiment_name` | `rough_direct` |
| `max_iterations` | 8000 |
| `save_interval` | 50 |
| `num_steps_per_env` | 24 |
| Policy | MLP `[128,128,128]`, ELU, `ActorCritic`, `init_noise_std` 1.0 |
| Algorithm | PPO: `learning_rate` 1e-3, **adaptive** schedule, `gamma` 0.99, `lam` 0.95, `clip_param` 0.2, `entropy_coef` 0.005, `num_learning_epochs` 5, `num_mini_batches` 4, `desired_kl` 0.01 |
| `empirical_normalization` | False |

### 9) Not enabled by default in Blind (`B2FlatEnvCfg` defaults)

| Flag | Default |
|------|---------|
| `use_imu` | False |
| `use_concurrent_state_est` | False |
| `use_rma` | False |
| `use_asymmetric_ppo` | False |
| `use_amp` | False |
| Terrain curriculum (`terrain_generator.curriculum`) | **False** for Rough-Blind cfg |

### 10) Deploy (MuJoCo) — `deploy/`

| Item | Detail |
|------|--------|
| Env vars | `BASIC_LOCOMOTION_ROBOT=b2`; **`BASIC_LOCOMOTION_POLICY_DIR`** = training run dir with **`params/env.yaml`** + **`exported/policy.onnx`** (not `.../exported` alone); typo **`BASIC_LOCOMMOTION_*`** accepted. Optional **`BASIC_LOCOMOTION_SCENE`**: default **`flat`** for b2 (`flat` / `random_boxes`). |
| `play_mujoco.py` | `sim_dt` **0.002** s; velocity command **`human`** (keyboard; need \|v cmd\| ≳ 0.01); outer torque **PD**: `tau = Kp*(q_des−q) − Kd*qdot`. Policy called every `round(1/(RL_FREQ*sim_dt))` steps; **`RL_FREQ = 1/(env.yaml sim.dt × decimation)`**. |
| Deploy PD (`deploy/config.py`, `b2`) | **`Kp_walking` 20**, **`Kd_walking` 1.5** (differs from Isaac asset **25 / 2** — common sim2sim tuning point). Stand keys **`Kp_stand_up_and_down` 25**, **`Kd_stand_up_and_down` 2** loaded in wrapper; **`play_mujoco.py`** uses walking gains only. |
| Isaac train vs deploy | No rough terrain / no domain randomization in MuJoCo; no obs noise by default; separate PD layer — see §3 vs §10 for gain mismatch. |
| Troubleshooting | Same task Blind vs Vision must match yaml/onnx; run Isaac **`play.py`** once to export ONNX; focus viewer + arrow keys; **`Space`** unpause. |

```bash
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.126.09             Driver Version: 580.126.09     CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 4060 ...    Off |   00000000:01:00.0 Off |                  N/A |
| N/A   38C    P8              2W /   55W |      13MiB /   8188MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A           10706      G   /usr/bin/gnome-shell                      2MiB |
+-----------------------------------------------------------------------------------------+
```

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/cbdb1a1a-b144-458a-bbb3-c93e8c243436" />
<img width="1430" height="483" alt="Image" src="https://github.com/user-attachments/assets/06b02884-e8b7-4c45-ad2a-b52cf50b67b8" />

**Comment by giulioturrisi** [comment]
Hello! B2 is the only one that we did not train at all. I will have a look, to me code-wise seems that everything should be fine. Let me check in the next few days.

**Comment by giulioturrisi** [comment]
Hi, just increased the gain of b2 and it works. Pushed on main. Note that the gains should not be high, otherwise you have a superstiff policy. You need to find a good compromise.

https://github.com/user-attachments/assets/532867b3-857f-447f-ae56-9817b768ca28

**Comment by Siwakonrome** [comment]
@giulioturrisi 
Thank you very much for checking this and for updating the gain for the B2. I really appreciate your time, effort, and guidance.

Thanks again for your support and for taking the time to look into this.
